README.md
### 빌드 환경
* ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=423f791d8a7fd6c80919698fa0f9e214248472f5, for GNU/Linux 2.6.32, stripped


### 설명
* 이 프로젝트는 F-code와 Secret key를 사용하여 로봇의 인증을 가능하게 합니다. 일단 인증되면 로봇 장치 정보는 추가 처리 및 제어를 위해 서버로 안전하게 전송됩니다.
* This project enables authentication of a robot using its F-code and secret key. Once authenticated, the robot device information is securely transmitted to the server for further processing and control.

### 실행방법
아래는 flyinglet-device-authentication 다운로드하고 등록하는 절차입니다.
<br/><br/>
  
1. 터미널을 열고 다음 명령어를 입력하여 git을 설치합니다.
```bash
sudo apt-get install git 
```
<br/><br/>
  
2. 파일을 다운로드 받을 디렉토리로 이동합니다.
```bash
cd /path/to/download/dir
```
<br/><br/>
  
3. git을 사용하여 flyinglet-device-authentication 을 다운로드 합니다.
```bash
git clone https://github.com/IN-TEGRIT/flyinglet-device-authentication.git
```
<br/><br/>
  
4. 다운로드된 flyinglet-device-authentication 디렉토리로 이동합니다.
```bash
cd flyinglet-device-authentication
```
<br/><br/>
  
5. flyinglet-device-authentication 에 실행 권한을 부여합니다.
```bash
chmod +x ./flyinglet-device-authentication
```
<br/><br/>
    
6. flyinglet-device-authentication 을 실행합니다.
```bash
./flyinglet-device-authentication
```
<br/><br/>

7. 실행 후, 등록하려는 f-code와 secret key를 입력합니다.
<br/><br/>
  
등록이 완료되면 flyinglet을 사용할 수 있습니다.



history.md
v1.0.0
- 인증 기능 구현

v1.1.0
- 네트워크 연결 실패시 재시도 기능 추가

v1.2.0 (23.09.14)
- 버전 확인 args 추가 (--version)

v1.3.0 (23.09.22)
- Log 기능 추가

v1.3.1 (23.09.26)
- Log 경로 변경 (log 폴더 내 저장)
- check-device 함수 requests.post의 verify=False


build_test.py
import curses
import curses.ascii
import uuid
import requests
import os
import argparse
import time
import datetime
import platform
import socket
from dotenv import set_key


class InstallIntegrit:
    def __init__(self):
        self.mac_address = None
        self.getnode = None
        self.machine = None
        self.node = None
        self.platform = None
        self.processor = None
        self.release = None
        self.system = None
        self.version = None
        self.uname = None
        self.authentication_timestamp = None
        self.env_path = os.path.expanduser('~/.flyinglet/.env')
        self.env_vars = {}
        self.airpath_path = os.path.expanduser('~/persist/data.sysinfo')
        self.env_vars = {}

        if os.path.exists(self.env_path):
            # .env 파일이 존재할 경우, 해당 파일을 로드
            with open(self.env_path) as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith('#') and len(line.split('=', 1)) == 2:
                        key, value = line.split('=', 1)
                        os.environ[key] = value
        else:
            # .env 파일이 존재하지 않을 경우, 생성
            os.makedirs(os.path.dirname(self.env_path), exist_ok=True)
            with open(self.env_path, 'w') as f:
                f.write('# Add your environment variables here\n')


        try:
        # airpath_info.json JSON 파일 읽기
            with open(self.airpath_path, 'r', encoding='UTF-8') as file:
                # 파일 내용
                file_contents = file.read()
                # 파일 내용을 문자열로 디코딩
                file_contents_str = file_contents
                # 줄 바꿈 문자로 분할한 후 키-값 쌍 추출 및 딕셔너리로 저장
                data = {}
                lines = file_contents_str.split('\n')   # 윈도우: \\n  우분투: \n
                for line in lines:
                    if ':' in line:
                        key, value = line.split(':', 1)  # 첫 번째 ':'만 기준으로 분리
                        data[key.strip()] = value.strip()

                # 딕셔너리 출력
                print(data)

        except:
            data = {}

        # 변환된 데이터 사용
        self.sn = data.get('MB', None)
        self.som_sn = data.get('SOM', None)
        self.ethernet_mac = data.get('Ethernet Mac',None)
        self.wifi_mac = data.get('WIFI Mac', None)
        self.bt_mac = data.get('BT Mac', None)
        self.name = data.get('MODEL', None)
        self.version = data.get('HW', None)

    def check_network_connection(self, timeout=3, retry_count=10):
        # 연결 상태 확인할 호스트와 포트
        host = "www.google.com"
        port = 80
        for i in range(retry_count):
            try:
                socket.setdefaulttimeout(timeout)
                socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
                return True
            except socket.error as ex:
                time.sleep(1)  # 1초 대기 후 재시도

        return False


    def Boot_routine(self):
        # Get sn - fcode
        sn_registry = self.get_sn_registry()
        if sn_registry is False or self.check_data(self.sn) is False:
            return print(False)
        else:
            fcode = sn_registry['fCode']
            if (os.environ.get('FCODE','').strip('\"\'') != fcode or
                    os.environ.get('AUTHENTICATION_TIMESTAMP','').strip('\"\'') != fcode or
                    sn_registry['mac_address'] != self.mac_address):
                self.get_mac_address()
                # Send to MongoDB
                now = datetime.datetime.now()
                unix_time = int(time.mktime(now.timetuple()))
                self.authentication_timestamp = unix_time

                # F-code 환경변수에 저장
                set_key(self.env_path, 'AUTHENTICATION_TIMESTAMP', str(self.authentication_timestamp))
                set_key(self.env_path, 'FCODE', str(fcode))
                self.send_macaddress(fcode)
            else:
                pass


    def argpaser(self):
        if self.check_network_connection(retry_count=10):
            # try:
            self.Boot_routine()
            # except Exception as e:
            #     print("An error occurred:", e)
        else:
            exit()

    def get_mac_address(self):
        # Get MAC address
        self.mac_address = ':'.join(
            ['{:02x}'.format((uuid.getnode() >> ele) & 0xff) for ele in range(0, 8 * 6, 8)][::-1]
        )
        self.getnode = uuid.getnode(),
        self.machine = platform.machine(),
        self.node = platform.node(),
        self.platform = platform.platform(),
        self.processor = platform.processor(),
        self.release = platform.release(),
        self.system = platform.system(),
        self.version = platform.version(),
        self.uname = platform.uname(),

    @staticmethod
    def registry_airpath(agent_access_key, agent_secret_key):
        urls = 'https://dev.api.flyinglet.com/agent/registry_airpath'

        datas = {"agent_access_key": agent_access_key, "agent_secret_key": agent_secret_key}
        user_id = requests.post(url=urls, data=datas)

        return user_id


    def get_sn_registry(self):
        urls = f'https://dev.api.flyinglet.com/fcode_registry?SN={self.sn}'
        try:
            sn_registry = requests.get(url=urls).json()[0]
            fcode = sn_registry['fCode']
            return sn_registry
        except:
            return False


    @staticmethod
    def check_data(sn):
        urls = f'https://dev.api.flyinglet.com/chip_sn?SN={sn}'
        response = requests.get(url=urls)
        return response.json()


    def send_macaddress(self,fcode):
        urls = f'https://dev.api.flyinglet.com/send-device-certi'
        datas = {"mac_address": self.mac_address,
                 "authentication_timestamp": self.authentication_timestamp,
                 "getnode": self.getnode,
                 "machine": self.machine,
                 "node": self.node,
                 "platform": self.platform,
                 "processor": self.processor,
                 "release": self.release,
                 "system": self.system,
                 "version": self.version,
                 "uname": self.uname,
                 "fCode": fcode
                 }
        requests.patch(url=urls, data=datas)

    def send_airpath(self, user_id):
        urls = f'https://dev.api.flyinglet.com/Airpath'
        datas = {"SN": self.sn,
                 "SOM S/N": self.som_sn,
                 "Ethernet Mac": self.ethernet_mac,
                 "WIFI Mac": self.wifi_mac,
                 "BT Mac": self.bt_mac,
                 "모델명": self.name,
                 "Version": self.version,

                 "user_id": user_id,
                 "status": "등록대기"
                 }
        requests.post(url=urls, data=datas)


    def verification(self, stdscr):
        # Init curses
        stdscr = curses.initscr()

        # No echo
        curses.noecho()

        # Squence key option
        stdscr.keypad(True)

        # Set maximum retry count
        max_retry_count = 3

        # Loop until user enters correct credentials or maximum retry count is reached
        retry_count = 0
        while retry_count < max_retry_count:

            # Clear screen
            stdscr.clear()

            # Print prompt to screen
            stdscr.addstr(0, 0, 'Enter AGENT_ACCESS_KEY: ')

            # Wait for user input
            agent_access_key = ""

            while True:
                c = stdscr.getch()
                if c == ord('\n'):
                    break
                elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                    if len(agent_access_key.rstrip()) > 0:
                        agent_access_key = agent_access_key[:-1]
                        stdscr.addstr(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key), ' ')  # 기존의 문자를 공백으로 덮어씀
                        stdscr.move(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key))
                    # else:
                    #     fCode = ''
                else:
                    # Print AGENT_ACCESS_KEY to screen
                    stdscr.move(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key))
                    agent_access_key += chr(c)
                    stdscr.addch(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key) - 1, c)

            # Print newline character
            stdscr.clear()

            # Print prompt to screen
            stdscr.addstr(0, 0, 'Enter AGENT_SECRET_KEY: ')

            # Hide AGENT_SECRET_KEY input
            agent_secret_key = ''

            # curses.noecho()
            while True:
                c = stdscr.getch()
                if c == ord('\n'):
                    break
                elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                    if len(agent_secret_key.rstrip()) > 0:
                        agent_secret_key = agent_secret_key[:-1]
                        stdscr.addstr(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key), ' ')  # 기존의 문자를 공백으로 덮어씀
                        stdscr.move(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key))
                    else:
                        agent_secret_key = ''
                        stdscr.move(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key))
                else:
                    agent_secret_key += chr(c)

                    # Print AGENT_SECRET_KEY to screen
                    stdscr.addch(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key) - 1, '*')

            response = self.registry_airpath(agent_access_key, agent_secret_key)

            # 아이디 유무 확인
            if response.status_code == 200:
                self.send_airpath(response.text)
                stdscr.clear()
                stdscr.addstr(0, 0, 'verification successfully completed. The program will now exit. Please press any key.')

                # Wait for user input
                stdscr.getch()
                break

            else:
                # Increment retry count
                retry_count += 1

                # Print error message to screen
                stdscr.clear()
                stdscr.addstr(0, 0, 'Incorrect AGENT_ACCESS_KEY or AGENT_SECRET_KEY. Retry {}/{}'.format(retry_count, max_retry_count))

                # Wait for user input
                stdscr.getch()
                stdscr.clear()

            # If maximum retry count is reached, print error message to screen and wait for user input
            if retry_count == max_retry_count:
                stdscr.addstr(0, 0, 'Maximum retry count reached. Please try again later.')
                stdscr.refresh()
                stdscr.getch()
                break




if '__main__' == __name__:
    parser = argparse.ArgumentParser()
    parser.add_argument('--check', action='store_true', help='Automatically checks device authentication')
    args = parser.parse_args()
    install_integrit = InstallIntegrit()
    if args.check:
        install_integrit.argpaser()
    else:
        # 입력 파일 인자가 없을 경우 기본 파일을 사용
        verification = install_integrit.verification
        curses.wrapper(verification)

flyinglet-device-authentication.py
    import curses
    import curses.ascii
    import uuid
    import requests
    import os
    import argparse
    import time
    import datetime
    import json
    import platform
    from dotenv import load_dotenv, set_key
    
    
    class InstallIntegrit:
        def __init__(self):
            self.mac_address = None
            self.getnode = None
            self.machine = None
            self.node = None
            self.platform = None
            self.processor = None
            self.release = None
            self.system = None
            self.version = None
            self.uname = None
            self.authentication_timestamp = None
            self.env_path = os.path.expanduser('~/.flyinglet/.env')
            self.env_vars = {}
            if os.path.exists(self.env_path):
                # .env 파일이 존재할 경우, 해당 파일을 로드
                with open(self.env_path) as f:
                    for line in f:
                        line = line.strip()
                        if line and not line.startswith('#'):
                            key, value = line.split('=', 1)
                            os.environ[key] = value
            else:
                # .env 파일이 존재하지 않을 경우, 생성
                os.makedirs(os.path.dirname(self.env_path), exist_ok=True)
                with open(self.env_path, 'w') as f:
                    f.write('# Add your environment variables here\n')
                print(f'{self.env_path} has been created.')
    
        def argpaser(self):
            try:
                saved_fcode = os.environ.get('FCODE').strip('\"\'')
                saved_authentication_timestamp = os.environ.get('AUTHENTICATION_TIMESTAMP').strip('\"\'')
                check_data = self.check_data(saved_fcode)
                if check_data[0]["authentication_timestamp"] == saved_authentication_timestamp:
                    pass
                else:  # authentication_timestamp 가 일치하지 않을 경우 .env 내용 삭제
                    with open(self.env_path, 'w') as f:
                        f.write('')
            except (Exception,):
                # authentication_timestamp 가 일치하지 않을 경우 .env 내용 삭제
                with open(self.env_path, 'w') as f:
                    f.write('')
    
        def get_mac_address(self):
            # Get MAC address
            self.mac_address = ':'.join(
                ['{:02x}'.format((uuid.getnode() >> ele) & 0xff) for ele in range(0, 8 * 6, 8)][::-1]
            )
            self.getnode = uuid.getnode(),
            self.machine = platform.machine(),
            self.node = platform.node(),
            self.platform = platform.platform(),
            self.processor = platform.processor(),
            self.release = platform.release(),
            self.system = platform.system(),
            self.version = platform.version(),
            self.uname = platform.uname(),
    
        @staticmethod
        def signin_device(fcode, secretkey):
            urls = 'https://dev.api.flyinglet.com/signin-device'
            datas = {"fCode": fcode, "secretKey": secretkey}
            response = requests.post(url=urls, data=datas)
            return response
    
        @staticmethod
        def check_data(fcode):
            urls = 'https://dev.api.flyinglet.com/check-device'
            datas = {"fCode": fcode}
            response = requests.post(url=urls, data=datas)
            return json.loads(response.text)
    
        def send_macaddress(self, fcode, secretkey):
            urls = f'https://dev.api.flyinglet.com/send-device-certi'
            datas = {"fCode": fcode,
                     "secretKey": secretkey,
                     "mac_address": self.mac_address,
                     "authentication_timestamp": self.authentication_timestamp,
                     "getnode": self.getnode,
                     "machine": self.machine,
                     "node": self.node,
                     "platform": self.platform,
                     "processor": self.processor,
                     "release": self.release,
                     "system": self.system,
                     "version": self.version,
                     "uname": self.uname,
                     }
            requests.patch(url=urls, data=datas)
    
        def verification(self, stdscr):
            # Init curses
            stdscr = curses.initscr()
    
            # No echo
            curses.noecho()
    
            # Squence key option
            stdscr.keypad(True)
    
            # Set maximum retry count
            max_retry_count = 3
    
            # Loop until user enters correct credentials or maximum retry count is reached
            retry_count = 0
            while retry_count < max_retry_count:
    
                # Clear screen
                stdscr.clear()
    
                # Print prompt to screen
                stdscr.addstr(0, 0, 'Enter F-CODE: ')
    
                # Wait for user input
                fcode = ""
    
                while True:
                    c = stdscr.getch()
                    if c == ord('\n'):
                        break
                    elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                        if len(fcode.rstrip()) > 0:
                            fcode = fcode[:-1]
                            stdscr.addstr(0, len('Enter F-CODE: ') + len(fcode), ' ')  # 기존의 문자를 공백으로 덮어씀
                            stdscr.move(0, len('Enter F-CODE: ') + len(fcode))
                        # else:
                        #     fCode = ''
                    else:
                        # Print F-CODE to screen
                        stdscr.move(0, len('Enter F-CODE: ') + len(fcode))
                        fcode += chr(c)
                        stdscr.addch(0, len('Enter F-CODE: ') + len(fcode)-1, c)
    
                # Print newline character
                stdscr.addstr(1, 0, '')
    
                # Print prompt to screen
                stdscr.addstr(2, 0, 'Enter SECRET_KEY: ')
    
                # Hide SECRET_KEY input
                secretkey = ''
                # curses.noecho()
                while True:
                    c = stdscr.getch()
                    if c == ord('\n'):
                        break
                    elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                        if len(secretkey.rstrip()) > 0:
                            secretkey = secretkey[:-1]
                            stdscr.addstr(2, len('Enter SECRET_KEY: ') + len(secretkey), ' ')  # 기존의 문자를 공백으로 덮어씀
                            stdscr.move(2, len('Enter SECRET_KEY: ') + len(secretkey))
                        else:
                            secretkey = ''
                            stdscr.move(2, len('Enter SECRET_KEY: ') + len(secretkey))
                    else:
                        secretkey += chr(c)
    
                        # Print SECRET_KEY to screen
                        stdscr.addch(2, len('Enter SECRET_KEY: ') + len(secretkey) - 1, '*')
    
                # stdscr.addstr(6, 0, f'Enter robot ID: {response}')
    
                response = install_integrit.signin_device(fcode, secretkey)
    
                # 아이디 유무 확인
                if response.status_code == 200:
                    stdscr.addstr(4, 0, 'verification successfully completed')
                    # Get MAC address
                    install_integrit.get_mac_address()
    
                    # 등록된 MAC 확인
                    load_data = install_integrit.check_data(fcode)
                    saved_mac = load_data[0]['mac_address']
                    saved_time = load_data[0]['authentication_timestamp']
                    if self.mac_address != saved_mac or os.environ.get('AUTHENTICATION_TIMESTAMP') != saved_time:
                        # question answer
                        question = 'Already registered F-code. Would you like to replace it with this robot? [y/n]: '
                        answer = ''
    
                        # Print newline character
                        stdscr.addstr(6, 0, '')
    
                        # Print prompt to screen
                        stdscr.addstr(6, 0, question)
    
                        # curses.noecho()
                        while True:
                            c = stdscr.getch()
                            if c == ord('\n'):
                                break
                            elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                                if len(answer.rstrip()) > 0:
                                    answer = answer[:-1]
                                    stdscr.addstr(6, len(question) + len(answer), ' ')  # 기존의 문자를 공백으로 덮어씀
                                    stdscr.move(6, len(question) + len(answer))
                                else:
                                    answer = ''
                                    stdscr.move(6, len(question) + len(answer))
                            else:
                                answer += chr(c)
    
                                # Print SECRET_KEY to screen
                                stdscr.addch(6, len(question) + len(answer) - 1, c)
    
                        if answer == "y":
                            # Send to MongoDB
                            now = datetime.datetime.now()
                            unix_time = int(time.mktime(now.timetuple()))
                            self.authentication_timestamp = unix_time
                            # F-code 환경변수에 저장
                            self.env_vars['AUTHENTICATION_TIMESTAMP'] = self.authentication_timestamp
                            set_key(self.env_path, 'AUTHENTICATION_TIMESTAMP', str(self.authentication_timestamp))
                            os.environ['AUTHENTICATION_TIMESTAMP'] = str(self.authentication_timestamp)
                            install_integrit.send_macaddress(fcode, secretkey)
    
                    # Print newline character
                    stdscr.addstr(8, 0, "The task has been completed. The program will now exit. Please press any key.")
    
                    # Wait for user input
                    stdscr.getch()
    
                    # F-code 환경변수에 저장
                    # 1. 파일저장
                    self.env_vars['FCODE'] = fcode
                    set_key(self.env_path, 'FCODE', fcode)
    
                    # 2. os레벨 저장
                    os.environ['FCODE'] = fcode
    
                    # Exit loop
                    break
                else:
                    # Increment retry count
                    retry_count += 1
    
                    # Print error message to screen
                    stdscr.addstr(8, 0, 'Incorrect F-CODE or SECRET_KEY. Retry {}/{}'.format(retry_count, max_retry_count))
    
                    # Wait for user input
                    stdscr.refresh()
                    stdscr.getch()
    
            # If maximum retry count is reached, print error message to screen and wait for user input
            if retry_count == max_retry_count:
                stdscr.addstr(8, 0, 'Maximum retry count reached. Please try again later.')
                stdscr.refresh()
                stdscr.getch()
    
    
    if '__main__' == __name__:
        parser = argparse.ArgumentParser()
        parser.add_argument('--check', action='store_true', help='Automatically checks device authentication')
        args = parser.parse_args()
        install_integrit = InstallIntegrit()
        if args.check:
            install_integrit.argpaser()
        else:
            # 입력 파일 인자가 없을 경우 기본 파일을 사용
            verification = install_integrit.verification
            curses.wrapper(verification)


flyinglet-airpath-authentication_aarch64.py
import curses
import curses.ascii
import uuid
import requests
import os
import argparse
import time
import datetime
import platform
import socket
from dotenv import set_key


class InstallIntegrit:
    def __init__(self):
        self.mac_address = None
        self.getnode = None
        self.machine = None
        self.node = None
        self.platform = None
        self.processor = None
        self.release = None
        self.system = None
        self.version = None
        self.uname = None
        self.authentication_timestamp = None
        self.env_path = os.path.expanduser('~/.flyinglet/.env')
        self.env_vars = {}
        self.airpath_path = os.path.expanduser('/persist/data/.sysinfo')
        self.env_vars = {}

        if os.path.exists(self.env_path):
            # .env 파일이 존재할 경우, 해당 파일을 로드
            with open(self.env_path) as f:
                for line in f:
                    line = line.strip()
                    if line and not line.startswith('#') and len(line.split('=', 1)) == 2:
                        key, value = line.split('=', 1)
                        os.environ[key] = value
        else:
            # .env 파일이 존재하지 않을 경우, 생성
            os.makedirs(os.path.dirname(self.env_path), exist_ok=True)
            with open(self.env_path, 'w') as f:
                f.write('# Add your environment variables here\n')


        # try:
        # airpath_info.json JSON 파일 읽기
        with open(self.airpath_path, 'r', encoding='UTF-8') as file:
            # 파일 내용
            file_contents = file.read()
            # 파일 내용을 문자열로 디코딩
            file_contents_str = file_contents
            # 줄 바꿈 문자로 분할한 후 키-값 쌍 추출 및 딕셔너리로 저장
            data = {}
            lines = file_contents_str.split('\n')   # 윈도우: \\n  우분투: \n
            for line in lines:
                if ':' in line:
                    key, value = line.split(':', 1)  # 첫 번째 ':'만 기준으로 분리
                    data[key.strip()] = value.strip()

            # 딕셔너리 출력
            print(data)

        # except:
        #     data = {}

        # 변환된 데이터 사용
        self.sn = data.get('MB', None)
        self.som_sn = data.get('SOM', None)
        self.name = data.get('MODEL', None)
        self.airpath_version = data.get('HW', None)
        self.sw = data.get('SW', None)

    def check_network_connection(self, timeout=3, retry_count=10):
        # 연결 상태 확인할 호스트와 포트
        host = "www.google.com"
        port = 80
        for i in range(retry_count):
            try:
                socket.setdefaulttimeout(timeout)
                socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
                return True
            except socket.error as ex:
                time.sleep(1)  # 1초 대기 후 재시도

        return False


    def Boot_routine(self):
        # Get sn - fcode
        sn_registry = self.get_sn_registry()
        if sn_registry is False or self.check_data(self.sn) is False:
            return print(False)
        else:
            fcode = sn_registry['fCode']
            if (os.environ.get('FCODE','').strip('\"\'') != fcode or
                    os.environ.get('AUTHENTICATION_TIMESTAMP','').strip('\"\'') != fcode or
                    sn_registry['mac_address'] != self.mac_address):
                self.get_mac_address()
                # Send to MongoDB
                now = datetime.datetime.now()
                unix_time = int(time.mktime(now.timetuple()))
                self.authentication_timestamp = unix_time

                # F-code 환경변수에 저장
                set_key(self.env_path, 'AUTHENTICATION_TIMESTAMP', str(self.authentication_timestamp))
                set_key(self.env_path, 'FCODE', str(fcode))
                self.send_macaddress(fcode)
            else:
                pass


    def argpaser(self):
        if self.check_network_connection(retry_count=10):
            # try:
            self.Boot_routine()
            # except Exception as e:
            #     print("An error occurred:", e)
        else:
            exit()

    def get_mac_address(self):
        # Get MAC address
        self.mac_address = ':'.join(
            ['{:02x}'.format((uuid.getnode() >> ele) & 0xff) for ele in range(0, 8 * 6, 8)][::-1]
        )
        self.getnode = uuid.getnode(),
        self.machine = platform.machine(),
        self.node = platform.node(),
        self.platform = platform.platform(),
        self.processor = platform.processor(),
        self.release = platform.release(),
        self.system = platform.system(),
        self.version = platform.version(),
        self.uname = platform.uname(),

    @staticmethod
    def registry_airpath(agent_access_key, agent_secret_key):
        urls = 'https://dev.api.flyinglet.com/agent/registry_airpath'

        datas = {"agent_access_key": agent_access_key, "agent_secret_key": agent_secret_key}
        user_id = requests.post(url=urls, data=datas)

        return user_id


    def get_sn_registry(self):
        urls = f'https://dev.api.flyinglet.com/fcode_registry?SN={self.sn}'
        try:
            sn_registry = requests.get(url=urls).json()[0]
            fcode = sn_registry['fCode']
            return sn_registry
        except:
            return False


    @staticmethod
    def check_data(sn):
        urls = f'https://dev.api.flyinglet.com/chip_sn?SN={sn}'
        response = requests.get(url=urls)
        return response.json()


    def send_macaddress(self,fcode):
        urls = f'https://dev.api.flyinglet.com/send-device-certi'
        datas = {"mac_address": self.mac_address,
                 "authentication_timestamp": self.authentication_timestamp,
                 "getnode": self.getnode,
                 "machine": self.machine,
                 "node": self.node,
                 "platform": self.platform,
                 "processor": self.processor,
                 "release": self.release,
                 "system": self.system,
                 "version": self.version,
                 "uname": self.uname,
                 "fCode": fcode
                 }
        requests.patch(url=urls, data=datas)

    def send_airpath(self, user_id):
        urls = f'https://dev.api.flyinglet.com/Airpath'
        datas = {"SN": self.sn,
                 "SOM S/N": self.som_sn,
                 "모델명": self.name,
                 "Version": self.airpath_version,
                 "SW": self.sw,

                 "user_id": user_id,
                 "status": "등록대기"
                 }
        requests.post(url=urls, data=datas)


    def verification(self, stdscr):
        # Init curses
        stdscr = curses.initscr()

        # No echo
        curses.noecho()

        # Squence key option
        stdscr.keypad(True)

        # Set maximum retry count
        max_retry_count = 3

        # Loop until user enters correct credentials or maximum retry count is reached
        retry_count = 0
        while retry_count < max_retry_count:

            # Clear screen
            stdscr.clear()

            # Print prompt to screen
            stdscr.addstr(0, 0, 'Enter AGENT_ACCESS_KEY: ')

            # Wait for user input
            agent_access_key = ""

            while True:
                c = stdscr.getch()
                if c == ord('\n'):
                    break
                elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                    if len(agent_access_key.rstrip()) > 0:
                        agent_access_key = agent_access_key[:-1]
                        stdscr.addstr(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key), ' ')  # 기존의 문자를 공백으로 덮어씀
                        stdscr.move(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key))
                    # else:
                    #     fCode = ''
                else:
                    # Print AGENT_ACCESS_KEY to screen
                    stdscr.move(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key))
                    agent_access_key += chr(c)
                    stdscr.addch(0, len('Enter AGENT_ACCESS_KEY: ') + len(agent_access_key) - 1, c)

            # Print newline character
            stdscr.clear()

            # Print prompt to screen
            stdscr.addstr(0, 0, 'Enter AGENT_SECRET_KEY: ')

            # Hide AGENT_SECRET_KEY input
            agent_secret_key = ''

            # curses.noecho()
            while True:
                c = stdscr.getch()
                if c == ord('\n'):
                    break
                elif c == curses.ascii.BS or c == curses.KEY_BACKSPACE:
                    if len(agent_secret_key.rstrip()) > 0:
                        agent_secret_key = agent_secret_key[:-1]
                        stdscr.addstr(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key), ' ')  # 기존의 문자를 공백으로 덮어씀
                        stdscr.move(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key))
                    else:
                        agent_secret_key = ''
                        stdscr.move(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key))
                else:
                    agent_secret_key += chr(c)

                    # Print AGENT_SECRET_KEY to screen
                    stdscr.addch(0, len('Enter AGENT_SECRET_KEY: ') + len(agent_secret_key) - 1, '*')

            response = self.registry_airpath(agent_access_key, agent_secret_key)

            # 아이디 유무 확인
            if response.status_code == 200:
                self.send_airpath(response.text)
                stdscr.clear()
                stdscr.addstr(0, 0, 'verification successfully completed. The program will now exit. Please press any key.')

                # Wait for user input
                stdscr.getch()
                break

            else:
                # Increment retry count
                retry_count += 1

                # Print error message to screen
                stdscr.clear()
                stdscr.addstr(0, 0, 'Incorrect AGENT_ACCESS_KEY or AGENT_SECRET_KEY. Retry {}/{}'.format(retry_count, max_retry_count))

                # Wait for user input
                stdscr.getch()
                stdscr.clear()

            # If maximum retry count is reached, print error message to screen and wait for user input
            if retry_count == max_retry_count:
                stdscr.addstr(0, 0, 'Maximum retry count reached. Please try again later.')
                stdscr.refresh()
                stdscr.getch()
                break




if '__main__' == __name__:
    parser = argparse.ArgumentParser()
    parser.add_argument('--check', action='store_true', help='Automatically checks device authentication')
    args = parser.parse_args()
    install_integrit = InstallIntegrit()
    if args.check:
        install_integrit.argpaser()
    else:
        # 입력 파일 인자가 없을 경우 기본 파일을 사용
        verification = install_integrit.verification
        curses.wrapper(verification)